content/en/blog/2022-06-24-transparently-mocking-grpc-services.md

```diff
@@ -0,0 +1,339 @@
+	}
+
+	jsonpb := &protojson.UnmarshalOptions{
+		AllowPartial:   true,
+		DiscardUnknown: true,
+	}
+
+	switch rpc {
+	case ScheduleRPC:
+		out := &api.Item{}
+		if err = jsonpb.Unmarshal(data, out); err != nil {
+			return fmt.Errorf("could not unmarshal json into %T: %v", out, err)
+		}
+		s.OnSchedule = func(context.Context, *api.Item) (*api.Item, error) {
+			return out, nil
+		}
+	case DailyRPC:
+		out := &api.Docket{}
+		if err = jsonpb.Unmarshal(data, out); err != nil {
+			return fmt.Errorf("could not unmarshal json into %T: %v", out, err)
+		}
+		s.OnDaily = func(context.Context, *api.Day) (*api.Docket, error) {
+			return out, nil
+		}
+	default:
+		return fmt.Errorf("unknown RPC %q", rpc)
+	}
+
+	return nil
+}
+
+// UseError allows you to specify a gRPC status error to return from the specified RPC.
+func (s *AgendaServer) UseError(rpc string, code codes.Code, msg string) error {
+	switch rpc {
+	case ScheduleRPC:
+		s.OnSchedule = func(context.Context, *api.Item) (*api.Item, error) {
+			return nil, status.Error(code, msg)
+		}
+	case DailyRPC:
+		s.OnDaily = func(context.Context, *api.Day) (*api.Docket, error) {
+			return nil, status.Error(code, msg)
+		}
+	default:
+		return fmt.Errorf("unknown RPC %q", rpc)
+	}
+	return nil
+}
+```
+
+## Putting It All Together
+
+With our mock package, writing gRPC tests becomes more of a matter of what we want to test rather than how to test it. The test below verifies that our mock works by testing an error path and a "happy" path.
+
+```go
+func TestSchedule(t *testing.T) {
+	// Setup AgendaServer mock
+	srv := mock.New(nil)
+	defer srv.Shutdown()
+
+	// Create a client to test that is connected to the mock server.
+	dialer := grpc.WithContextDialer(srv.Channel().Dialer)
+	creds := grpc.WithTransportCredentials(insecure.NewCredentials())
+	agenda, err := client.New("bufconn", dialer, creds)
+	require.NoError(t, err, "could not connect to remote agenda via bufconn")
+
+	// Test the case where the server returns an error
+	srv.UseError(mock.ScheduleRPC, codes.DataLoss, "something bad happened")
+	err = agenda.Schedule("hello", "world", time.Now(), time.Now().Add(5*time.Minute))
+	require.Error(t, err, "expected an error returned from the server")
+	require.Equal(t, 1, srv.Calls[mock.ScheduleRPC])
+
+	// Test the case where server returns a response
+	srv.UseFixture(mock.ScheduleRPC, "testdata/item.json")
+	err = agenda.Schedule("hello", "world", time.Now(), time.Now().Add(5*time.Minute))
+	require.NoError(t, err, "expected no error in happy path")
+	require.Equal(t, 2, srv.Calls[mock.ScheduleRPC])
+}
+```
+
+## Conclusion
+
+Mocking gRPC services for the purposes of testing isn't always straightforward. However, with tools like `bufconn` and some clever code factoring we can create an in-memory mock that is completely isolated from production code. Introducing some measure of configurability into the mocks can also help us focus on _what_ we're testing rather than _how_ we're testing it.
```

Comment:
minor suggestion
```suggestion
Mocking gRPC services for the purposes of testing isn't always straightforward. However, with tools like `bufconn` and some clever code factoring, we can create an in-memory mock that is completely isolated from production code. Introducing some measure of configurability into the mocks can also help us focus on _what_ we're testing rather than _how_ we're testing it.
```